prow/cmd/deck/main.go

```diff
@@ -1607,3 +1617,65 @@ func defaultLensRemoteConfig(lfc *config.LensFileConfig) error {
 
 	return nil
 }
+
+// validateStoragePath "validates" the storage bucket, *if* the `Deck.RestrictStoragePaths` config field is true.
+// The path should match a generic "<key-type>/<bucket>/<folder>/<sub-folder...>" pattern.
+// Additionally, the path's bucket should match either the built-in and default storage locations _or_
+// should be contained in the `Deck.AdditionalAllowedBuckets` config field.
+// If the `Deck.RestrictStoragePaths` field is false, no validation errors are returned (but there will be warnings logged).
+func validateStoragePath(cfg config.Getter, path string) error {
+	validate := func(cfg config.Getter, path string) error {
+		parts := strings.SplitN(path, "/", 4)
+		if len(parts) < 4 {
+			return httpError{
+				error:      fmt.Errorf("invalid path: %s expected <key-type>/<bucket>/<folder>/<sub-folder...>", path),
+				statusCode: http.StatusBadRequest,
+			}
+		}
+
+		// Check bucket
+		bucket := parts[1]
+		allowed_buckets := sets.NewString(cfg().Deck.AdditionalAllowedBuckets...)
+		for _, dc := range cfg().Plank.DefaultDecorationConfigs {
```

Comment:
There is another source of valid buckets we have to consider: They can be set on jobs. So we need to walk all jobs, check if they have `.DecorationConfig != nil && .DecorationConfig.GCSConfiguration != nil && .DecorationConfig.GCSConfiguration.BucketName != "`.

Maybe it would be better to construct that set in `Load` and store it somewhere on the config?

Reply:
Done.

_Note_: This can't handle buckets defined on jobs via InRepoConfig. For now (until ~Jan 2021), I'm skipping validation on InRepoConfig-enabled Prows and providing a detailed warning on how to prepare.